# Going through error by error to show mistakes.

In [5]:
import pandas as pd
import numpy as np
from sklearn import LinearRegression
from sklearn.cross_validation import cross_val_score

# Load data
d = pd.read_csv('train.csv')


# Setup data for prediction
x1 = data.SalaryNormalized
x2 = pd.get_dummies(data.ContractType)

# Setup model
model = LinearRegression()

# Evaluate model
from sklearn.cross_validation import cross_val_score
from sklearn.cross_validation import train_test_split
scores = cross_val_score(model, x2, x1, cv=1, scoring='mean_absolute_error')
print(scores.mean())

ImportError: cannot import name LinearRegression

In [10]:
#Error 1
# In order to correctly import the model, you need to specify the subdirectory of sklearn in this case.
from sklearn.linear_model import LinearRegression

In [6]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.cross_validation import cross_val_score

# Load data
d = pd.read_csv('train.csv')


# Setup data for prediction
x1 = data.SalaryNormalized
x2 = pd.get_dummies(data.ContractType)

# Setup model
model = LinearRegression()

# Evaluate model
from sklearn.cross_validation import cross_val_score
from sklearn.cross_validation import train_test_split
scores = cross_val_score(model, x2, x1, cv=1, scoring='mean_absolute_error')
print(scores.mean())

NameError: name 'data' is not defined

# Error 2
data is not defined. Python doesn't know what to do with it.


In [7]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.cross_validation import cross_val_score

# Load data
d = pd.read_csv('train.csv')


# Setup data for prediction
x1 = d.SalaryNormalized
x2 = pd.get_dummies(d.ContractType)

# Setup model
model = LinearRegression()

# Evaluate model
from sklearn.cross_validation import cross_val_score
from sklearn.cross_validation import train_test_split
scores = cross_val_score(model, x2, x1, cv=1, scoring='mean_absolute_error')
print(scores.mean())

ValueError: k-fold cross validation requires at least one train / test split by setting n_folds=2 or more, got n_folds=1.

# Error 3
k-fold cross validation works by comparing how an algorithm works on multiple folds, so cv has to be an integer greater than 1 in order for it to work. If you have only one sample, there is nothing to cross-check your model against.

In [12]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.cross_validation import cross_val_score

# Load data
d = pd.read_csv('train.csv')


# Setup data for prediction
x1 = d.SalaryNormalized
x2 = pd.get_dummies(d.ContractType)

# Setup model
model = LinearRegression()

# Evaluate model
from sklearn.cross_validation import cross_val_score
from sklearn.cross_validation import train_test_split
scores = cross_val_score(model, x2, x1, cv=4, scoring='mean_absolute_error')
print(scores.mean())

-11766.4002944


/home/matthew/anaconda2/lib/python2.7/site-packages/sklearn/metrics/scorer.py:90: DeprecationWarning: Scoring method mean_absolute_error was renamed to neg_mean_absolute_error in version 0.18 and will be removed in 0.20.
  sample_weight=sample_weight)
/home/matthew/anaconda2/lib/python2.7/site-packages/sklearn/metrics/scorer.py:90: DeprecationWarning: Scoring method mean_absolute_error was renamed to neg_mean_absolute_error in version 0.18 and will be removed in 0.20.
  sample_weight=sample_weight)
/home/matthew/anaconda2/lib/python2.7/site-packages/sklearn/metrics/scorer.py:90: DeprecationWarning: Scoring method mean_absolute_error was renamed to neg_mean_absolute_error in version 0.18 and will be removed in 0.20.
  sample_weight=sample_weight)
/home/matthew/anaconda2/lib/python2.7/site-packages/sklearn/metrics/scorer.py:90: DeprecationWarning: Scoring method mean_absolute_error was renamed to neg_mean_absolute_error in version 0.18 and will be removed in 0.20.
  sample_weight=sample_

#### This works, it tells us that our cross-validation score is -11766.4 when we have our set up correctly.

# Convenience Tip
We generally put our imports at the top of the page, so that everything is imported first, because it makes coding easier when you are writing a lot. It isn't wrong, it is just inconvenient. 

# Convention Tip
as is normal in mathematics, we generally name our functions y and X by convention. x1 and x2 are not wrong, but it makes it easier to read and will fit into formulas you find when you look for tools online.